### ISSUE TITLE
Fail to start server on windows with VSCode

### ISSUE THREAD

**Issue by Caestille** [apple/sourcekit-lsp#681]
I can't seem to start the sourcekit lsp server when using the VSCode swift extension from sswg. I have installed the windows 10 swift release and pointed the extension config to the swift binaries directory at C:\Program Files\swift\runtime-development\usr\bin and swift executables at C:\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin. The server initialisation fails without much info, tries 5 times and then gives up:

[Trace - 09:59:54] Sending request 'initialize - (0)'.
[Info  - 09:59:54] Connection to server got closed. Server will restart.
[Error - 09:59:54] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] SourceKit Language Server client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Trace - 09:59:54] Sending request 'initialize - (0)'.
[Info  - 09:59:54] Connection to server got closed. Server will restart.
[Error - 09:59:54] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] SourceKit Language Server client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Trace - 09:59:54] Sending request 'initialize - (0)'.
[Info  - 09:59:54] Connection to server got closed. Server will restart.
[Error - 09:59:54] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] SourceKit Language Server client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] Restarting server failed
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Trace - 09:59:54] Sending request 'initialize - (0)'.
[Info  - 09:59:54] Connection to server got closed. Server will restart.
[Error - 09:59:54] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] SourceKit Language Server client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] Restarting server failed
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Trace - 09:59:54] Sending request 'initialize - (0)'.
[Error - 09:59:54] The SourceKit Language Server server crashed 5 times in the last 3 minutes. The server will not be restarted. See the output for more information.
[Error - 09:59:54] Server initialization failed.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] SourceKit Language Server client: couldn't create connection to server.
  Message: Pending response rejected since connection got disposed
  Code: -32097 
[Error - 09:59:54] Restarting server failed
  Message: Pending response rejected since connection got disposed
  Code: -32097

With the lack of info I'm not sure what's going wrong, any idea what I might be missing?

**Comment by ahoppen** [comment]
rdar://103510110

**Comment by adam-fowler** [comment]
@Caestille what version of swift do you have installed? Run `swift --version`. 

Also can you run `swift build` outside of VSCode without issue?

**Comment by compnerd** [comment]
I think that this might be a packaging issue :-(.  I cannot run the tests on Windows anymore because LSP is no longer reasonably sized.

**Comment by compnerd** [comment]
@Caestille assuming that https://dev.azure.com/compnerd/swift-build/_build/results?buildId=61082&view=results passes, would you mind testing with the resulting build?  If that works, it was a packaging issue, with SK-LSP growing a dependency on SwiftSyntax, but that was being built dynamically and not being packaged for re-distribution.

**Comment by compnerd** [comment]
https://artprodeus21.artifacts.visualstudio.com/A8fd008a0-56bc-482c-ba46-67f9425510be/3133d6ab-80a8-4996-ac4f-03df25cd3224/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2NvbXBuZXJkL3Byb2plY3RJZC8zMTMzZDZhYi04MGE4LTQ5OTYtYWM0Zi0wM2RmMjVjZDMyMjQvYnVpbGRJZC82MTA4Mi9hcnRpZmFjdE5hbWUvaW5zdGFsbGVyLWFtZDY00/content?format=file&subPath=%2Finstaller.exe is the one-off installer.  It seems to work locally, I'm closing this issue for now, feel free to re-open if this doesn't work for you.  I'm going to kick off another build for the 12/19 snapshot.  Sorry about the breakage!

**Comment by Caestille** [comment]
> @Caestille what version of swift do you have installed? Run `swift --version`.
> 
> Also can you run `swift build` outside of VSCode without issue?

Hi, sorry for the delay: 
![image](https://user-images.githubusercontent.com/50808413/209109853-dc648987-af36-4706-852d-a738bf332f1b.png)

I do get 'error: toolchain is invalid: could not find link' when trying to build a newly initialised executable, could this be the issue, or am I missing an argument in my build command?

**Comment by Caestille** [comment]
> https://artprodeus21.artifacts.visualstudio.com/A8fd008a0-56bc-482c-ba46-67f9425510be/3133d6ab-80a8-4996-ac4f-03df25cd3224/_apis/artifact/cGlwZWxpbmVhcnRpZmFjdDovL2NvbXBuZXJkL3Byb2plY3RJZC8zMTMzZDZhYi04MGE4LTQ5OTYtYWM0Zi0wM2RmMjVjZDMyMjQvYnVpbGRJZC82MTA4Mi9hcnRpZmFjdE5hbWUvaW5zdGFsbGVyLWFtZDY00/content?format=file&subPath=%2Finstaller.exe is the one-off installer. It seems to work locally, I'm closing this issue for now, feel free to re-open if this doesn't work for you. I'm going to kick off another build for the 12/19 snapshot. Sorry about the breakage!

I'll give this a go and let you know. Thanks!

**Comment by Caestille** [comment]
> @Caestille assuming that https://dev.azure.com/compnerd/swift-build/_build/results?buildId=61082&view=results passes, would you mind testing with the resulting build? If that works, it was a packaging issue, with SK-LSP growing a dependency on SwiftSyntax, but that was being built dynamically and not being packaged for re-distribution.

Unfortunately I get the same error and server failing to start

**Comment by compnerd** [comment]
You are not running the build in the correct shell.  The Visual Studio Developer Command Prompt must be used or you need to source the vsdevenv.

**Comment by Caestille** [comment]
Ah thanks, sorry about that. I get this with this set of commands:

C:\Users\jward\OneDrive\Documents\Hello>swift package init --type executable
Creating executable package: Hello
Creating Package.swift
Creating README.md
Creating .gitignore
Creating Sources/
Creating Sources\Hello\Hello.swift
Creating Tests/
Creating Tests\HelloTests/
Creating Tests\HelloTests\HelloTests.swift

C:\Users\jward\OneDrive\Documents\Hello>swift build
error: 'hello': Invalid manifest (compiled with: ["C:\\Library\\Developer\\Toolchains\\unknown-Asserts-development.xctoolchain\\usr\\bin\\swiftc.exe", "-L", "C:\\Library\\Developer\\Toolchains\\unknown-Asserts-development.xctoolchain\\usr\\lib\\swift\\pm\\ManifestAPI", "-lPackageDescription", "-sdk", "C:\\Library\\Developer\\Platforms\\Windows.platform\\Developer\\SDKs\\Windows.sdk", "-libc", "MD", "-I", "C:\\Library\\Developer\\Platforms\\Windows.platform\\Developer\\Library\\XCTest-development\\usr\\lib\\swift\\windows", "-I", "C:\\Library\\Developer\\Platforms\\Windows.platform\\Developer\\Library\\XCTest-development\\usr\\lib\\swift\\windows\\x86_64", "-L", "C:\\Library\\Developer\\Platforms\\Windows.platform\\Developer\\Library\\XCTest-development\\usr\\lib\\swift\\windows\\x86_64", "-use-ld=lld", "-swift-version", "5", "-I", "C:\\Library\\Developer\\Toolchains\\unknown-Asserts-development.xctoolchain\\usr\\lib\\swift\\pm\\ManifestAPI", "-package-description-version", "5.8.0", "C:\\Users\\jward\\AppData\\Local\\Temp\\TemporaryDirectory.V8Njvw\\manifest.swift", "-Xfrontend", "-disable-implicit-concurrency-module-import", "-Xfrontend", "-disable-implicit-string-processing-module-import", "-o", "C:\\Users\\jward\\AppData\\Local\\Temp\\TemporaryDirectory.J0mTEU\\hello-manifest.exe"])
error: link command failed with exit code 1 (use -v to see invocation)
lld-link: error: <root>: undefined symbol: mainCRTStartup
lld-link: error: undefined symbol: memcpy
>>> referenced by C:\Users\jward\AppData\Local\Temp\TemporaryDirectory.kpzKIF\manifest-1.o:($s18PackageDescription6TargetC10DependencyOWOb)
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by compnerd** [comment]
It’s still the same problem.  You’re not using the correct shell.  Make sure that you’re using the Native Developer Command Primot or are sites Int VsDevEnv with the correct settings.

**Comment by Caestille** [comment]
I'm using the VS2019 developer command prompt, is that what you meant? 
![image](https://user-images.githubusercontent.com/50808413/209155067-6245503b-750b-467e-b731-0234c8f78f56.png)

**Comment by compnerd** [comment]
No, that is not the correct developer command prompt.  That is the x86 version, which is why you are seeing the undefined symbols.  Ensure that you are using the x64 one (which is the Native Developer Command Prompt).

**Comment by Caestille** [comment]
Ah ok. Sorry for the confusion, the build worked in there, so I assume the swift installation is fine? However the sourcekit lsp server is still failing to start in VS code

**Comment by compnerd** [comment]
Did you run `code .` from the VS Developer Command Prompt?

**Comment by Caestille** [comment]
Tried that just now. Still same error in VSCode output for the sourcekit language server

**Comment by compnerd** [comment]
Okay; and what happens if you run `sourcekit-lsp` from the Native Developer Command Prompt?

**Comment by Caestille** [comment]
Calling it with a debug log level I get an 'unsupported platform' warning 5 times and then nothing

**Comment by compnerd** [comment]
That sounds like it is launching fine from the command line, the question is now why does it fail to launch from VSCode?  Try `set SOURCEKIT_LOGGING=3` before launching `code .` (and make sure to enable the verbose logging in VSCode's settings for SourceKit-LSP).  That might help identify what is going on.

**Comment by Caestille** [comment]
Ah, I tried setting the log level to verbose (was debug) and I saw an error indicating it didn't like the format of the arguments (despite simply copying the example in the VSCode settings :/ ). I removed the arguments and the server seems able to start and function in VSCode (I'm getting autocompletion suggestions etc). I'm not sure why I didn't spot the error before. Sorry to have wasted your time with what turned out to be nothing :/
